### DISCUSSION TITLE
Extracting the Channel Impulse Response from the gNB

### DISCUSSION THREAD

**Discussion by melkael** [srsran/srsRAN_Project#804] [General Help] [answered]
Hello,

I am working on a project which requires to extract the channel impulse response from the gNB. Is there a way to do that? Or could you please direct me to the portion of the code that I would have to modify if necessary?

Thank you

**Comment by dvdgrgrtt** [comment, marked as answer]
Hi @melkael,

if you have access to MATLAB, the easiest way is to use one of [our analyzers](https://github.com/srsran/srsRAN_matlab#appsanalyzerssrspuschanalyzer). Basically, you tell the gNB to record all uplink samples (organized in resource grid), then you can use the analyzers to look into these records - and one of the things they plot is the estimated channel response (in frequency, not in time). For more info, have a loot at [this tutorial](https://docs.srsran.com/projects/project/en/latest/tutorials/source/matlab/source/index.html).

If you don't have access to MATLAB, you can still record the uplink resource grids and then process them manually (I suggest you look into the MATLAB functions to understand how to extract a relevant section of samples from the trace file). The problem here is that these are the samples received by the gNB, and the channel needs to be estimated.

The last option, as you say, is to modify the code - what I'd do is probably print the content of `ch_estimate` after you run the estimator (for instance, [here](https://github.com/srsran/srsRAN_Project/blob/4ac5300d4927b5199af69e6bc2e55d061fc33652/lib/phy/upper/channel_processors/pusch/pusch_processor_impl.cpp#L180). Something like `fmt::print("{}", ch_estimate.get_symbol_ch_estimate(2, 0))` should output to console the frequency response of symbol 2 (in the current slot), for antenna port 0. Of course, printing (or recording) everything will slow the gNB down...

Hope this helps!
